### PR TITLE
Make phaser stateless

### PR DESCRIPTION
**Pull request by pietro68**
Resolves: #3689 

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/effects/Phaser.h [hunk 1]
```diff
@@ -91,21 +91,21 @@ class EffectPhaser final : public StatefulPerTrackEffect
...
+      return 0;
+   }
+
+   // Temporarily needed
+   size_t RealtimeProcess(size_t group, EffectSettings& settings,
+      const float* const* inbuf, float* const* outbuf, size_t numSamples)
+      override
+   {
+      return 0;
+   }
```
Comment (Paul-Licameli) [inline src/effects/Phaser.h]: I don't like the unexplained "Temporarily needed" functions.

Instead, change the base class to PerTrackEffect, then you don't need these two overrides.

Also, declare `MakeInstance` as override.

File: src/effects/Phaser.cpp [hunk 2]
```diff
@@ -130,6 +130,61 @@ struct EffectPhaser::Validator
...
+      float* const* outBlock,
+      size_t                 blockLen);
+
+   void Coefficients(double hz, double slope, double gain, double samplerate, int type,
+      double& a0, double& a1, double& a2, double& b0, double& b1, double& b2);
+
+   float DoFilter(EffectPhaserState& data, float in);
+
+   EffectPhaserState mMaster;
+   std::vector<EffectPhaserState> mSlaves;
```
Comment (Paul-Licameli) [inline src/effects/Phaser.cpp:176]: Commenting on the whole PR, it appears to be a correct transformation.  But I want to add a requirement.

Make the objects in this array the same class as the instance -- as is done already for AudioUnits and VST2.

When that works, do the same when you transform other built-in effects to statelessness.  Write PRs to do the same retroactively for the Reverb and Wahwah transformations.

If time allows, make PRs to do the same for Ladspa and LV2.

The reason?

I don't like the duplicated interface for processing EffectInterface.h.  There is RealtimeProcess but also ProcessBlock.  I wabt them unified so the interface is simpler.

That means eliminating the group argument of realtime process.  More about that later:  but making every array of slaves an array of Instances is the beginning.  Then the "array of slaves" is exposed to RealtimeEffectState, not an internal detail of each Instance class.  The RealtimeEffectState will dispatch to the correct Instance object.

File: src/effects/Phaser.h (hunk 1 above)
Comment (pietro68) [reply src/effects/Phaser.h]: Ok, integrated that in step 3

Comment (pietro68) [reply src/effects/Phaser.h]: Ok, both integrated into step 3 now.

**Review by Paul-Licameli — APPROVED** [review]
All looks good.

Adapt the rest of the real-time capable built-ins to this same pattern in other pull requests.

Those may soon include Distortion and Echo.

**Pull request merged**